File: backend/enum-jpa-validation-app/src/main/java/com/example/enumapp/common/code/CodeEnumJacksonConfig.java
package com.example.enumapp.common.code;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.BeanDescription;
import com.fasterxml.jackson.databind.DeserializationConfig;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.deser.BeanDeserializerModifier;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.fasterxml.jackson.databind.ser.BeanSerializerModifier;
import org.springframework.boot.autoconfigure.jackson.Jackson2ObjectMapperBuilderCustomizer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.io.IOException;

/**
 * Request/Response JSON 에서 CodeEnum 을 code 문자열로 직렬화/역직렬화한다.
 * null, "", "   " 는 null 로 변환한다.
 */
@Configuration
public class CodeEnumJacksonConfig {

    @Bean
    Jackson2ObjectMapperBuilderCustomizer codeEnumJacksonCustomizer() {
        return builder -> {
            SimpleModule module = new SimpleModule("CodeEnumModule");
            module.setDeserializerModifier(new BeanDeserializerModifier() {
                @Override
                public JsonDeserializer<?> modifyEnumDeserializer(
                        DeserializationConfig config,
                        com.fasterxml.jackson.databind.JavaType type,
                        BeanDescription beanDesc,
                        JsonDeserializer<?> deserializer
                ) {
                    Class<?> raw = type.getRawClass();
                    if (raw.isEnum() && CodeEnum.class.isAssignableFrom(raw)) {
                        @SuppressWarnings({"unchecked", "rawtypes"})
                        Class<? extends Enum> enumClass = (Class<? extends Enum>) raw;
                        return new CodeEnumDeserializer(enumClass);
                    }
                    return deserializer;
                }
            });
            module.setSerializerModifier(new BeanSerializerModifier() {
                @Override
                public JsonSerializer<?> modifyEnumSerializer(
                        com.fasterxml.jackson.databind.SerializationConfig config,
                        com.fasterxml.jackson.databind.JavaType valueType,
                        BeanDescription beanDesc,
                        JsonSerializer<?> serializer
                ) {
                    if (CodeEnum.class.isAssignableFrom(valueType.getRawClass())) {
                        return new CodeEnumSerializer();
                    }
                    return serializer;
                }
            });
            builder.modulesToInstall(module);
        };
    }

    static final class CodeEnumSerializer extends JsonSerializer<CodeEnum> {
        @Override
        public void serialize(CodeEnum value, JsonGenerator gen, SerializerProvider serializers) throws IOException {
            if (value == null) {
                gen.writeNull();
            } else {
                gen.writeString(value.getCode());
            }
        }
    }

    static final class CodeEnumDeserializer<E extends Enum<E> & CodeEnum> extends JsonDeserializer<E> {
        private final Class<E> type;

        CodeEnumDeserializer(Class<E> type) {
            this.type = type;
        }

        @Override
        public E deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
            String code = p.getValueAsString();
            return CodeEnums.fromCode(type, code);
        }

        @Override
        public E getNullValue(DeserializationContext ctxt) {
            return null;
        }
    }
}

File: backend/enum-jpa-querydsl-kotlin-app/src/main/kotlin/com/example/enumapp/common/code/CodeEnumJacksonConfig.kt
package com.example.enumapp.common.code

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.BeanDescription
import com.fasterxml.jackson.databind.DeserializationConfig
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.JsonSerializer
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.deser.BeanDeserializerModifier
import com.fasterxml.jackson.databind.module.SimpleModule
import com.fasterxml.jackson.databind.ser.BeanSerializerModifier
import org.springframework.boot.autoconfigure.jackson.Jackson2ObjectMapperBuilderCustomizer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

/**
 * Request/Response JSON 에서 CodeEnum 을 code 문자열로 직렬화/역직렬화한다.
 * null, "", "   " 는 null 로 변환한다.
 */
@Configuration
class CodeEnumJacksonConfig {

    @Bean
    fun codeEnumJacksonCustomizer(): Jackson2ObjectMapperBuilderCustomizer =
        Jackson2ObjectMapperBuilderCustomizer { builder ->
            val module = SimpleModule("CodeEnumModule")
            module.setDeserializerModifier(object : BeanDeserializerModifier() {
                override fun modifyEnumDeserializer(
                    config: DeserializationConfig,
                    type: com.fasterxml.jackson.databind.JavaType,
                    beanDesc: BeanDescription,
                    deserializer: JsonDeserializer<*>
                ): JsonDeserializer<*> {
                    val raw = type.rawClass
                    if (raw.isEnum && CodeEnum::class.java.isAssignableFrom(raw)) {
                        return CodeEnumDeserializer(raw)
                    }
                    return deserializer
                }
            })
            module.setSerializerModifier(object : BeanSerializerModifier() {
                override fun modifyEnumSerializer(
                    config: com.fasterxml.jackson.databind.SerializationConfig,
                    valueType: com.fasterxml.jackson.databind.JavaType,
                    beanDesc: BeanDescription,
                    serializer: JsonSerializer<*>
                ): JsonSerializer<*> {
                    if (CodeEnum::class.java.isAssignableFrom(valueType.rawClass)) {
                        return CodeEnumSerializer()
                    }
                    return serializer
                }
            })
            builder.modulesToInstall(module)
        }

    internal class CodeEnumSerializer : JsonSerializer<CodeEnum>() {
        override fun serialize(value: CodeEnum?, gen: JsonGenerator, serializers: SerializerProvider) {
            if (value == null) {
                gen.writeNull()
            } else {
                gen.writeString(value.code)
            }
        }
    }

    /**
     * Java 의 raw-type deserializer 와 동일하게 Class<*> 를 받는다.
     * (Kotlin 은 Enum&lt;E&gt; & CodeEnum 교차 제네릭을 호출부에 추론하기 어렵다)
     */
    internal class CodeEnumDeserializer(
        private val type: Class<*>
    ) : JsonDeserializer<Any?>() {

        override fun deserialize(p: JsonParser, ctxt: DeserializationContext): Any? {
            val code = p.valueAsString
            return fromCodeUnchecked(type, code)
        }

        override fun getNullValue(ctxt: DeserializationContext): Any? = null

        @Suppress("UNCHECKED_CAST")
        private fun fromCodeUnchecked(type: Class<*>, code: String?): Any? {
            if (code.isNullOrBlank()) {
                return null
            }
            val constants = type.enumConstants as Array<out CodeEnum>
            return constants.firstOrNull { it.code == code }
                ?: throw IllegalArgumentException(
                    "Unknown code '%s' for enum %s".format(code, type.simpleName)
                )
        }
    }
}
